python/cudf_polars/cudf_polars/experimental/expressions.py

```diff
@@ -0,0 +1,485 @@
+    exprs: Sequence[Expr | NamedExpr],
+    child_ir_count: int,
+    *,
+    update: MutableMapping[Expr, int] | None = None,
+    skip_fused_exprs: bool = False,
+) -> MutableMapping[Expr, int]:
+    """
+    Extract a partition-count mapping for Expr nodes.
+
+    Parameters
+    ----------
+    exprs
+        Sequence of root expressions to traverse and
+        get partition counts.
+    child_ir_count
+        Partition count for the child-IR node.
+    update
+        Existing mapping to update.
+    skip_fused_exprs
+        Whether to skip over FusedExpr objects. This
+        can be used to stay within a local FusedExpr
+        sub-expression.
+
+    Returns
+    -------
+    Mapping between Expr nodes and partition counts.
+    """
+    expr_partition_counts: MutableMapping[Expr, int] = update or {}
+    for e in exprs:
+        expr = e.value if isinstance(e, NamedExpr) else e
+        for node in list(traversal([expr]))[::-1]:
+            if isinstance(node, FusedExpr):
+                # Process the fused sub-expression graph first
+                if skip_fused_exprs:
+                    continue  # Stay within the current sub expression
+                expr_partition_counts = extract_partition_counts(
+                    [node.sub_expr],
+                    child_ir_count,
+                    update=expr_partition_counts,
+                    skip_fused_exprs=True,
+                )
+                expr_partition_counts[node] = expr_partition_counts[node.sub_expr]
+            elif isinstance(node, (Agg, Literal)):
+                # Assume all aggregations produce 1 partition
+                expr_partition_counts[node] = 1
+            elif node.is_pointwise:
+                # Pointwise expressions should preserve child partition count
+                if node.children:
+                    # Assume maximum child partition count
+                    expr_partition_counts[node] = max(
+                        [expr_partition_counts[c] for c in node.children]
+                    )
+                else:
+                    # If no children, we are preserving the child-IR partition count
+                    expr_partition_counts[node] = child_ir_count
+            else:  # pragma: no cover
+                raise NotImplementedError(
+                    f"{type(node)} not supported for multiple partitions."
+                )
+
+    return expr_partition_counts
+
+
+def _replace(e: Expr, rec: ExprTransformer) -> Expr:
+    return rec.state["mapping"].get(e, reuse_if_unchanged(e, rec))
+
+
+def replace(e: Expr, mapping: Mapping[Expr, Expr]) -> Expr:
+    """Replace one or more expression nodes."""
+    mapper = CachingVisitor(_replace, state={"mapping": mapping})
+    return mapper(e)
+
+
+def rename_agg(agg: Expr, new_name: str, *, new_options: Any = None) -> Expr:
+    """Modify the name of an aggregation expression."""
+    return replace(agg, {agg: Agg(agg.dtype, new_name, new_options, *agg.children)})
```

Comment:
I think there's more traversal than necessary here, why not just:

```
return Agg(agg.dtype, new_name, new_options, *agg.children)
```

Also, not sure if `agg.dtype` is correct as the dtype. For example, if we're "renaming" from `sum` to `count`, then `sum` could be an arbitrary type, but `count` is definitely an integer type.

Reply:
>I think there's more traversal than necessary here, why not just:

Sorry - You are absolutely right. We no longer need any of the `rename_agg`/`replace` utilities.

>Also, not sure if agg.dtype is correct as the dtype. For example, if we're "renaming" from sum to count, then sum could be an arbitrary type, but count is definitely an integer type.

Was also unsure about this. Is there a specific int type you'd like to use for all `"count"` aggregations?

Reply:
So libcudf will always produce int32. In a partitioned setting that may not be sufficient. So perhaps always produce uint64?